pkg/table/tables/partition.go

```diff
@@ -1957,53 +2039,91 @@ func partitionedTableUpdateRecord(ctx table.MutateContext, txn kv.Transaction, t
 				return errors.Trace(err)
 			}
 		}
-		memBuffer.Release(sh)
-		return nil
+		return finishFunc(err, nil)
 	}
-	tbl := t.getPartition(to)
-	err = tbl.updateRecord(ctx, txn, h, currData, newData, touched, opt)
-	if err != nil {
-		return errors.Trace(err)
+
+	var found map[string][]byte
+	var newFromKey, newToKey kv.Key
+
+	keys := make([]kv.Key, 0, 2)
+	encodedRecordID := codec.EncodeInt(nil, h.IntValue())
+	if newFrom != 0 {
+		newFromKey = tablecodec.EncodeRowKey(newFrom, encodedRecordID)
+		keys = append(keys, newFromKey)
 	}
-	if _, ok := t.reorganizePartitions[to]; ok {
-		// Even if to == from, in the reorganized partitions they may differ
-		// like in case of a split
-		newTo, err := t.locateReorgPartition(ectx.GetEvalCtx(), newData)
-		if err != nil {
-			return errors.Trace(err)
+	if !deleteOnly && newTo != 0 {
+		// Only need to check if writing.
+		if newTo == newFrom {
+			newToKey = newFromKey
+		} else if newRecordHandle.Equal(h) {
+			// And no new record id generated (else new unique id, cannot be found)
+			newToKey = tablecodec.EncodeRowKey(newTo, encodedRecordID)
+			keys = append(keys, newToKey)
 		}
-		newFrom, err := t.locateReorgPartition(ectx.GetEvalCtx(), currData)
+	}
+	if len(keys) > 0 {
+		found, err = txn.BatchGet(context.Background(), keys)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if newTo == newFrom {
-			tbl = t.getPartition(newTo)
-			if deleteOnly {
-				err = tbl.RemoveRecord(ctx, txn, h, currData)
-			} else {
-				err = tbl.updateRecord(ctx, txn, h, currData, newData, touched, opt)
-			}
+	}
+	var newToKeyAndValIsSame *bool
+	if newFrom != 0 {
+		if val, ok := found[string(newFromKey)]; ok {
+			var same bool
+			same, err = dataEqRec(ctx.GetExprCtx().GetEvalCtx().Location(), t.Meta(), currData, val)
 			if err != nil {
 				return errors.Trace(err)
 			}
-			memBuffer.Release(sh)
-			return nil
+			if same {
+				// Always do Remove+Add, to always have the indexes in-sync,
+				// since the indexes might not been created yet, i.e. not backfilled yet.
+				err = t.getPartition(newFrom).RemoveRecord(ctx, txn, h, currData)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			}
+			if newTo == newFrom {
+				newToKeyAndValIsSame = &same
+			}
 		}
-		tbl = t.getPartition(newFrom)
-		err = tbl.RemoveRecord(ctx, txn, h, currData)
-		if err != nil {
-			return errors.Trace(err)
+	}
+	if deleteOnly || newTo == 0 {
+		return finishFunc(err, nil)
+	}
+	if val, ok := found[string(newToKey)]; ok {
+		if newToKeyAndValIsSame == nil {
+			same, err := dataEqRec(ctx.GetExprCtx().GetEvalCtx().Location(), t.Meta(), currData, val)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			newToKeyAndValIsSame = &same
 		}
-		if !deleteOnly {
-			tbl = t.getPartition(newTo)
-			_, err = tbl.addRecord(ctx, txn, newData, opt.GetAddRecordOpt())
+		if !*newToKeyAndValIsSame {
+			// Generate a new ID
+			newRecordHandle, err = AllocHandle(context.Background(), ctx, t)
```

Comment:
I have a question if we didn't alloc new handle id here just use the perious one, what will happen?

Reply:
Then the record in the other partition set (reorg set for <= StateWriteReorganization, original set for >= `StateDeleteReorganization`) would be overwritten. First case would mean a lost row, second case a glitch where an older transaction/tidb connection would temporarily not see old row.